added draft version of doc for networks with state (#3372)
openvinotoolkit/openvino: docs/IE_DG/network_state_intro.md

> Introduction to OpenVINO state API {#openvino_docs_IE_DG_network_state_intro}

Reviewer: we need to add this document to `ie_docs.xml` layout file. Otherwise, it's not a part of documentation.

                <edge from-layer="1" from-port="1" to-layer="3" to-port="0"/>
                <edge from-layer="3" from-port="2" to-layer="4" to-port="0"/>
                <edge from-layer="3" from-port="2" to-layer="10" to-port="0"/> 
                <edge from-layer="1" from-port="1" to-layer="10" to-port="1"/>
                <edge from-layer="10" from-port="2" to-layer="5" to-port="0"/>
	</edges>
	<meta_data>
		<MO_version value="unknown version"/>
		<cli_parameters>
		</cli_parameters>
	</meta_data>
</net>
```

### Example of creating model Ngraph API

```cpp
    auto arg = make_shared<op::Parameter>(element::f32, Shape{1, 1});
    auto init_const = op::Constant::create(element::f32, Shape{1, 1}, {0});
    auto read = make_shared<op::ReadValue>(init_const, "v0");
    std::vector<shared_ptr<Node>> args = {arg, read};
    auto add = make_shared<op::Add>(arg, read);
    auto assign = make_shared<op::Assign>(add, "v0");
    auto add2 = make_shared<op::Add>(add, read);
    auto res = make_shared<op::Result>(add2);

    auto f = make_shared<Function>(ResultVector({res}), ParameterVector({arg}), SinkVector({assign}));
```

> In this example `SinkVector` was used to create Ngraph::Function. For network with states except inputs and outputs also Assign nodes should be pointed to Function 

Reviewer: `Ngraph::Function` -> `ngraph::Function`. In this case doxygen will generate a link to `ngraph::Function` documentation

Author: fixed, but doxygen have not generated link. should I do something special to have such links (not only here but for other functions included in text)?

Reviewer: ok, let's keep as is and fix later. 

Now we need only green CI.

to avoid deleting it during graph transformations. It can be done with constructor as shown in example or with special method `Fuction::add_sink`. Also you can delete 
sink from `Function` after deleting node from graph with `Function::delete_sink` method.

## OpenVINO state API

 InferenceEngine have method `InferRequest::QueryState` to get list of states from network and IVariableState interface to operate with state. Brief description of methods 
 is below and next section contains small workable example how this interface can be used.
 
 * `std::string GetName() const`
   returns name(variable_id) of according Variable
 * `void Reset()`
   reset state to default value
 * `void SetState(Blob::Ptr newState)`
   set new value for state
 * `Blob::CPtr GetState() const`
   returns current value of state

## Example of inference network with state

Let's take IR from example in prevous sections. In this example inference of 2 independent sequences of data will be demonstrated. Between these sequences state 
should be reset.

> @snippet openvino/docs/snippets/InferenceEngine_network_with_state_infer.cpp part1

Reviewer: `openvino/docs/` should be removed. 

You can compile the documentation using `make openvino_docs` and make sure that your document looks well.

More powerfull examples of work with networks with states are sample and demo demonstrating work with speech. 

> Decsriptions can be found in [Samples Overview][{./Samples_Overview.md}

Reviewer: does  `[{./Samples_Overview.md}` work?

 between data portions should be addressed. For that networks saved some data between inferences - state. When one dependent sequence is over,
 state should be reset to initial value and new sequence can be started.
 
 Several frameworks have special API for states in networks. For example, Keras have special option for RNNs `stateful` that turn on saving state 
 between inferences. Kaldi contains special specifier `Offset` to define time offset in network. 
 
 OpenVINO contains special API to simplify work with networks with states too. State will be automatically saved between inferences 
 and there is a way to reset state when needed. Also user can read state or set it to some new value between inferences.
 
## OpenVINO state representation

 OpenVINO contains special abstraction Variable to represent state in network. There are 2 operations to work with state: 
* `Assign` to save value in state
* `ReadValue` to read value saved on previous iteration.

More details on these operations you can find in [specification](../ops/opset3.md).

## Examples of representation of network with states

> To get model with states ready for inference you can convert model from another framework with ModelOptimizer to IR or create it with Ngraph builder. 

Reviewer 2: "Ngraph builder"? Maybe it's better to say "to create ngraph function" and add a link to https://docs.openvinotoolkit.org/latest/openvino_docs_nGraph_DG_build_function.html

Author: fixed

Let's represent in both forms the following graph.

> [state_network_example]: img/state_network_example.png

Reviewer 2: looks like the png file is empty, could you please check it? (the screen is attached)
![image](https://user-images.githubusercontent.com/21181928/100735612-9166f000-33e2-11eb-86ad-320e3169f366.png)

Author: fixed

		</layer>
		<layer id="4" name="save" type="Assign" version="opset3">
			<data variable_id="id"/>
			<input>
				<port id="0">
					<dim>1</dim>
					<dim>4</dim>
				</port>
			</input>
		</layer>
		<layer id="5" name="output/sink_port_0" type="Result" version="opset1">
			<input>
				<port id="0">
					<dim>1</dim>
					<dim>4</dim>
				</port>
			</input>
		</layer>
	</layers>
	<edges>
		<edge from-layer="0" from-port="1" to-layer="1" to-port="0"/>
        <edge from-layer="2" from-port="0" to-layer="3" to-port="1"/>
        <edge from-layer="1" from-port="1" to-layer="6" to-port="0"/>
        <edge from-layer="7" from-port="1" to-layer="6" to-port="1"/>
        <edge from-layer="8" from-port="1" to-layer="6" to-port="2"/>
        <edge from-layer="9" from-port="1" to-layer="6" to-port="3"/>
        <edge from-layer="6" from-port="4" to-layer="3" to-port="0"/>
        <edge from-layer="3" from-port="2" to-layer="4" to-port="0"/>
        <edge from-layer="3" from-port="2" to-layer="5" to-port="0"/>
	</edges>
	<meta_data>
		<MO_version value="unknown version"/>
		<cli_parameters>
		</cli_parameters>
	</meta_data>
</net>
```

### Example of creating model Ngraph API

```cpp

>     auto arg = make_shared<op::Parameter>(element::f32, Shape{1, 4});

Author: fixed

are represented in IR and Ngraph, how operations with state can be done. Section provides small examples 
of stateful network and code to infer it.

## What is stateful network

 Several use cases require processing of data sequences. When length of sequence is known and small enough, 
 we can process it with RNN like networks which contains cycle inside. But in some cases like online speech recognition of time series 
 forecasting length of data sequence is unknown. Then data can be divided in small portions and processed step by step. But dependency 
 between data portions should be addressed. For that networks saved some data between inferences - state. When one dependent sequence is over,
 state should be reset to initial value and new sequence can be started.
 
 Several frameworks have special API for states in networks. For example, Keras have special option for RNNs `stateful` that turn on saving state 
 between inferences. Kaldi contains special specifier `Offset` to define time offset in network. 
 
 OpenVINO contains special API to simplify work with networks with states too. State will be automatically saved between inferences 

>  and there is a way to reset state when needed. Also user can read state or set it to some new value between inferences.

Reviewer 3: ```suggestion
 and there is a way to reset state when needed. You can also read state or set it to some new value between inferences.
```

 
## OpenVINO state representation

>  OpenVINO contains special abstraction Variable to represent state in network. There are 2 operations to work with state: 

Reviewer 3: ```suggestion
 OpenVINO contains special abstraction variable to represent state in a network. There are two operations to work with state: 
```

* `Assign` to save value in state

> * `ReadValue` to read value saved on previous iteration.

Reviewer 3: ```suggestion
* `ReadValue` to read value saved on previous iteration
```

> More details on these operations you can find in [ReadValue specification](../ops/infrastructure/ReadValue_3.md) and 

Reviewer 3: ```suggestion
You can find more details on these operations in [ReadValue specification](../ops/infrastructure/ReadValue_3.md) and 
```

[Assign specification](../ops/infrastructure/Assign_3.md).

> ## Examples of representation of network with states

Reviewer 3: ```suggestion
## Examples of representation of a network with states
```

> To get model with states ready for inference you can convert model from another framework with ModelOptimizer to IR or create ngraph function 

Reviewer 3: ```suggestion
To get a model with states ready for inference, you can convert a model from another framework to IR with Model Optimizer or create an nGraph function 
```

(details can be found in [Build nGraph Function section](../nGraph_DG/build_function.md)). 

> Let's represent in both forms the following graph:

Reviewer 3: ```suggestion
Let's represent the following graph in both forms:
```

![state_network_example]

### Example of IR with state

> `bin` file for this graph should contain float 0 in binary form. `xml` is the following.

Reviewer 3: ```suggestion
The `bin` file for this graph should contain float 0 in binary form. Content of `xml` is the following.
```

			<data axis="1"/>
			<input>
				<port id="0">
					<dim>1</dim>
					<dim>1</dim>
				</port>
				<port id="1">
					<dim>1</dim>
					<dim>1</dim>
				</port>
			</input>
			<output>
				<port id="2" precision="FP32">
					<dim>1</dim>
					<dim>1</dim>
				</port>
			</output>
		</layer>
		<layer id="5" name="output/sink_port_0" type="Result" version="opset5">
			<input>
				<port id="0">
					<dim>1</dim>
					<dim>1</dim>
				</port>
			</input>
		</layer>
	</layers>
	<edges>
		<edge from-layer="0" from-port="1" to-layer="1" to-port="0"/>
                <edge from-layer="2" from-port="0" to-layer="3" to-port="1"/>
                <edge from-layer="1" from-port="1" to-layer="3" to-port="0"/>
                <edge from-layer="3" from-port="2" to-layer="4" to-port="0"/>
                <edge from-layer="3" from-port="2" to-layer="10" to-port="0"/> 
                <edge from-layer="1" from-port="1" to-layer="10" to-port="1"/>
                <edge from-layer="10" from-port="2" to-layer="5" to-port="0"/>
	</edges>
	<meta_data>
		<MO_version value="unknown version"/>
		<cli_parameters>
		</cli_parameters>
	</meta_data>
</net>
```

> ### Example of creating model Ngraph API

Reviewer 3: ```suggestion
### Example of creating model nGraph API
```

```cpp
    auto arg = make_shared<op::Parameter>(element::f32, Shape{1, 1});
    auto init_const = op::Constant::create(element::f32, Shape{1, 1}, {0});
    auto read = make_shared<op::ReadValue>(init_const, "v0");
    std::vector<shared_ptr<Node>> args = {arg, read};
    auto add = make_shared<op::Add>(arg, read);
    auto assign = make_shared<op::Assign>(add, "v0");
    auto add2 = make_shared<op::Add>(add, read);
    auto res = make_shared<op::Result>(add2);

    auto f = make_shared<Function>(ResultVector({res}), ParameterVector({arg}), SinkVector({assign}));
```

> In this example `SinkVector` was used to create `ngraph::Function`. For network with states except inputs and outputs also Assign nodes should be pointed to Function 

Reviewer 3: ```suggestion
In this example, `SinkVector` is used to create `ngraph::Function`. For network with states, except inputs and outputs,  `Assign` nodes should also point to `Function` 
```

> to avoid deleting it during graph transformations. It can be done with constructor as shown in example or with special method `add_sinks(const SinkVector& sinks)`. Also you can delete 

Reviewer 3: ```suggestion
to avoid deleting it during graph transformations. You can do it with the constructor, as shown in the example, or with the special method `add_sinks(const SinkVector& sinks)`. Also you can delete 
```

> sink from `ngraph::Function` after deleting node from graph with `delete_sink()` method.

Reviewer 3: ```suggestion
sink from `ngraph::Function` after deleting the node from graph with the `delete_sink()` method.
```

## OpenVINO state API

>  InferenceEngine have method `InferRequest::QueryState` to get list of states from network and IVariableState interface to operate with state. Brief description of methods 

Reviewer 3: ```suggestion
 Inference Engine has the `InferRequest::QueryState` method  to get the list of states from a network and `IVariableState` interface to operate with states. Below you can find brief description of methods and the workable example of how to use this interface.  
```

 is below and next section contains small workable example how this interface can be used.
 
 * `std::string GetName() const`
   returns name(variable_id) of according Variable
 * `void Reset()`
   reset state to default value
 * `void SetState(Blob::Ptr newState)`
   set new value for state
 * `Blob::CPtr GetState() const`
   returns current value of state

> ## Example of inference stateful network

Reviewer 3: ```suggestion
## Example of stateful network inference
```

> Let's take IR from example in prevous sections. In this example inference of 2 independent sequences of data will be demonstrated. Between these sequences state should be reset.

Reviewer 3: ```suggestion
Let's take an IR from the previous section example. The example below demonstrates inference of two independent sequences of data. State should be reset between these sequences.
```

One infer request and one thread 
will be used in this example. Using several threads is possible if you have several independent sequences. Then each sequence can be processed in its own infer 
request. Inference of one sequence in several infer requests is not recommended. In one infer request state will be saved automatically between inferences, but 

> if the first step done in one infer request and the second in another, state should be set in new infer request manually (using `IVariableState::SetState` method).

Reviewer 3: ```suggestion
if the first step is done in one infer request and the second in another, state should be set in new infer request manually (using `IVariableState::SetState` method).
```

@snippet openvino/docs/snippets/InferenceEngine_network_with_state_infer.cpp part1

> More powerfull examples of work with networks with states are sample and demo demonstrating work with speech. 

Reviewer 3: ```suggestion
You can find more powerful examples demonstrating how to work with networks with states in speech sample and demo. 
```

Decsriptions can be found in [Samples Overview](./Samples_Overview.md)

[state_network_example]: ./img/state_network_example.png

## LowLatency transformation

> If the original framework does not have a special API for working with states, then after importing the model, OpenVINO representation will also not contain Assign/ReadValue layers. For example, if original ONNX model contain RNN operations, IR will contain TensorIterator operations and the values will be obtained only after the execution of whole TensorIterator primitive, intermediate values from each iteration will not be available. To be able to work with these intermediate values of each iteration and receive them with a low latency after each infer request, a special LowLatency transformation was introduced.

Reviewer 3: ```suggestion
If the original framework does not have a special API for working with states, after importing the model, OpenVINO representation will not contain Assign/ReadValue layers. For example, if the original ONNX model contains RNN operations, IR will contain TensorIterator operations and the values will be obtained only after the execution of whole TensorIterator primitive, intermediate values from each iteration will not be available. To be able to work with these intermediate values of each iteration and receive them with a low latency after each infer request, a special LowLatency transformation was introduced.
```

> LowLatency transformation changes the structure of the network containing [TensorIterator](../ops/infrastructure/TensorIterator_1.md) by adding the ability to work with state, inserting Assign/ReadValue layers as it shown on the picture below.

Reviewer 3: ```suggestion
LowLatency transformation changes the structure of the network containing [TensorIterator](../ops/infrastructure/TensorIterator_1.md) by adding the ability to work with state, inserting Assign/ReadValue layers as it is shown in the picture below.
```

> [applying_low_latency_example]: ./img/applying_low_latency.png

Reviewer 2: I will check. As I can see, it's also for line 198 in your part.

> [applying_low_latency_example]: ./img/applying_low_latency.png

Reviewer 2: fixed

> ## Steps to apply LowLatency transformation

Reviewer 2: done

1. Get CNNNetwork. Any way is acceptable:

	* [from IR or ONNX model](Integrate_with_customer_application_new_API.md#integration-steps)
	* [from nGraph Function](../nGraph_DG/build_function.md)

2. [Reshape](ShapeInference) CNNNetwork network if necessary 

> 	**Necessary case:** the sequence_lengths dimention of input > 1, it means TensorIterator layer will have number_iterations > 1. We should reshape the inputs of the network to set sequence_dimension exactly to 1.

Reviewer 3: dimension?

 
	```cpp

	// Network before reshape: Parameter (name: X, shape: [2 (sequence_lengths), 1, 16]) -> TensorIterator (num_iteration = 2, axis = 0) -> ...

	cnnNetwork.reshape({“X” : {1, 1, 16});

	// Network after reshape: Parameter (name: X, shape: [1 (sequence_lengths), 1, 16]) -> TensorIterator (num_iteration = 1, axis = 0) -> ...
		
	```
3. Apply LowLatency transformation

	```cpp
	#include "ie_transformations.hpp"
	
	...

	InferenceEngine::LowLatency(cnnNetwork);
	```

> 	**State naming rule:**  a name of state is a concatanation of names original TensorIterator operation, Parameter of the body and additional suffix "variable_" + id (0-base indexing, new indexing for each TensorIterator), e.g:

Reviewer 3: ```suggestion
	**State naming rule:**  a name of state is a concatenation of names: original TensorIterator operation, Parameter of the body, and additional suffix "variable_" + id (0-base indexing, new indexing for each TensorIterator), for example:
```

		tensor_iterator_name = "TI_name"
		body_parameter_name = "param_name"

		state_name = "TI_name/param_name/variable_0"

4. [Use state API](#openvino-state-api)

 
### Known limitations
1. Parameters are directly connected to States (ReadValues).

> 	Removing Parameters from `ngraph::Function` are not possible.

Reviewer 3: ```suggestion
	Removing Parameters from `ngraph::Function` is not possible.
```

	![low_latency_limitation_1](./img/low_latency_limitation_1.png)

	**Current solution:** replace Parameter with Constant (freeze) with the value [0, 0, 0 … 0] via [ModelOptimizer CLI](../MO_DG/prepare_model/convert_model/Converting_Model_General.md) `--input` or `--freeze_placeholder_with_value`.

2.  Non-reshapable network.

> 	Value of shapes is hardcoded somewhere in the network. 

Reviewer 3: ```suggestion
	Value of shapes is hard-coded somewhere in the network. 
```
